File: algorithms/medium/find_the_prefix_common_array_of_two_arrays.ex
defmodule Solution do
  @spec find_the_prefix_common_array(a :: [integer], b :: [integer]) :: [integer]
  def find_the_prefix_common_array(a, b) do
    Enum.zip(a, b)
    |> Enum.reduce({[], 0, %MapSet{}, %MapSet{}}, fn {a, b}, {ans, ct, sa, sb} ->
      ct =
        cond do
          a == b -> ct + 1
          MapSet.member?(sa, b) && MapSet.member?(sb, a) -> ct + 2
          MapSet.member?(sa, b) || MapSet.member?(sb, a) -> ct + 1
          true -> ct
        end

      {[ct | ans], ct, MapSet.put(sa, a), MapSet.put(sb, b)}
    end)
    |> elem(0)
    |> Enum.reverse()
  end
end

File: algorithms/medium/find_the_prefix_common_array_of_two_arrays.rs
use std::collections::HashSet;

impl Solution {
    pub fn find_the_prefix_common_array(a: Vec<i32>, b: Vec<i32>) -> Vec<i32> {
        let mut sa = HashSet::new();
        let mut sb = HashSet::new();

        let mut ans = Vec::new();
        let mut ct = 0;
        for i in 0..a.len() {
            if a[i] == b[i] {
                ct += 1;
            } else {
                if sa.contains(&b[i]) {
                    ct += 1;
                }
                if sb.contains(&a[i]) {
                    ct += 1;
                }
            }
            sa.insert(a[i]);
            sb.insert(b[i]);

            ans.push(ct);
        }
        ans
    }
}
